### ISSUE TITLE
SilverStripe 4.1.x and Silverstripe-S3 module - TinyMCE WYSIWYG doesn't load on initial CMS Load

### ISSUE THREAD

**Issue by kinglozzer** [silverstripe/silverstripe-s3#18]
_From @obj63mc on May 29, 2018 21:32_

Currently testing using v4.1.x and dev/master of silverstripe-s3.

I wasn't sure where the best place to log this issue but assuming it will probably need to be moved to a fix in silverstripe/framework as it is an issue with the HTMLEditorField, or possibly needs to go to the silverstripe-s3 repo.

# Steps to reproduce
1. Create a clean install of SS4.x using the composer command - 

            composer create-project silverstripe/installer /path/to/project 4.1.1
2. Configure the site as normal and require silverstripe-s3 

            composer require silverstripe/silverstripe-s3

3. Complete configuration of S3
4. Go to your site/admin and login
5. Edit the Home Page-  I get the following JS error - 

        bundle.js?m=1520903238:1 Uncaught ReferenceError: tinymce is not defined
            at Object.create (bundle.js?m=1520903238:1)
            at Object.init (bundle.js?m=1520903238:1)
            at init.onadd (bundle.js?m=1520903238:1)
            at o (vendor.js?m=1520903238:1)
            at Array.<anonymous> (vendor.js?m=1520903238:1)
            at HTMLDocument.<anonymous> (vendor.js?m=1520903238:1)
            at HTMLDocument.dispatch (eval at t.exports (vendor.js?m=1520903238:1), <anonymous>:3332:9)
            at HTMLDocument.eventHandle (eval at t.exports (vendor.js?m=1520903238:1), <anonymous>:2941:28)
            at Object.trigger (eval at t.exports (vendor.js?m=1520903238:1), <anonymous>:3210:12)
            at init.triggerHandler (eval at t.exports (vendor.js?m=1520903238:1), <anonymous>:3874:24)

Screen shot - 
![screen shot 2018-05-29 at 4 23 58 pm](https://user-images.githubusercontent.com/443120/40686377-b9425afa-635c-11e8-874f-51fac9677307.png)

I notice though that in my network console it is trying to load the tinymce script from the amazon public directory, not the local server address - example : 

        ....s3.amazonaws.com/public/_tinymce/tinymce-cms-92563982d3.js?_=1527628916905 

vs 
        /assets/_tinymce/tinymce-cms-92563982d3.js?_=1527628916905

Since this is loading the tinymce script file from S3 not sure if that is the cause of a race condition, cross domain origin issue or ?

If I refresh the page in the browser (eg /admin/pages/edit/show/1), the CMS loads normally and the WYSIWYG loads normally.  If I go to a different section of the CMS besides pages though such as say 'Settings' and then back to pages I get the same issue until I reload the site.

Is this possibly an issue as the file is being generated into the /public/assets directory instead of the /public/resources directory?  If there is a way to exclude this file from loading on S3 or a way to change the write path for the tinymce js I can test this some more to try and confirm what exactly is causing the issue.

_Copied from original issue: silverstripe/silverstripe-cms#2177_

**Comment by kinglozzer** [comment]
_From @maxime-rainville on July 13, 2018 3:18_

The `silverstripe/s3` module is experimental. It's not something we're actively working on at this time.

**Comment by kinglozzer** [comment]
_From @obj63mc on July 13, 2018 16:12_

As a note, I turned off and removed the S3 module and still have the same issue.  So not 100% sure it is caused by that module.

When trying to determine the cause of this.  My site is setup with the following - 

1. the 'mysite' code lives in the 'app' folder in the root.  I have a structure that is based off of - https://github.com/a2nt/silverstripe-webpack
2. The following modules are installed - 

            aws/aws-sdk-php                            3.62.11           
            colymba/gridfield-bulk-editing-tools       3.0.0-beta4       
            composer/ca-bundle                         1.1.1             
            composer/installers                        v1.5.0            
            dnadesign/silverstripe-elemental           2.1.0             
            dnadesign/silverstripe-elemental-list      1.0.0             
            dnadesign/silverstripe-elemental-userforms 1.0.0             
            dnadesign/silverstripe-elemental-virtual   dev-master 3867957
            doctrine/instantiator                      1.1.0             
            embed/embed                                v3.3.5            
            guzzlehttp/guzzle                          6.3.3             
            guzzlehttp/promises                        v1.3.1            
            guzzlehttp/psr7                            1.4.2             
            intervention/image                         2.4.2             
            jonom/silverstripe-betternavigator         4.1.1             
            league/flysystem                           1.0.45            
            league/flysystem-aws-s3-v3                 1.0.19            
            m1/env                                     2.1.2             
            marcj/topsort                              1.1.0             
            monolog/monolog                            1.23.0            
            moosylvania/silverstripe-age-gate          2.1.0             
            mtdowling/jmespath.php                     2.4.0             
            myclabs/deep-copy                          1.8.1             
            nikic/php-parser                           v3.1.5            
            paragonie/random_compat                    v2.0.17           
            phpdocumentor/reflection-common            1.0.1             
            phpdocumentor/reflection-docblock          4.3.0             
            phpdocumentor/type-resolver                0.4.0             
            phpspec/prophecy                           1.7.6             
            phpunit/php-code-coverage                  4.0.8             
            phpunit/php-file-iterator                  1.4.5             
            phpunit/php-text-template                  1.2.1             
            phpunit/php-timer                          1.0.9             
            phpunit/php-token-stream                   2.0.2             
            phpunit/phpunit                            5.7.27            
            phpunit/phpunit-mock-objects               3.4.4             
            psr/cache                                  1.0.1             
            psr/container                              1.0.0             
            psr/http-message                           1.0.1             
            psr/log                                    1.0.2             
            psr/simple-cache                           1.0.1             
            sebastian/code-unit-reverse-lookup         1.0.1             
            sebastian/comparator                       1.2.4             
            sebastian/diff                             1.4.3             
            sebastian/environment                      2.0.0             
            sebastian/exporter                         2.0.0             
            sebastian/global-state                     1.1.1             
            sebastian/object-enumerator                2.0.1             
            sebastian/recursion-context                2.0.0             
            sebastian/resource-operations              1.0.0             
            sebastian/version                          2.0.1             
            sheadawson/silverstripe-linkable           2.0.0             
            silverstripe/admin                         1.1.2             
            silverstripe/asset-admin                   1.1.2             
            silverstripe/assets                        1.1.2             
            silverstripe/blog                          3.2.0             
            silverstripe/campaign-admin                1.1.2             
            silverstripe/cms                           4.1.2             
            silverstripe/comment-notifications         2.0.1             
            silverstripe/comments                      3.1.2             
            silverstripe/config                        1.0.4             
            silverstripe/elemental-blocks              1.0.0             
            silverstripe/errorpage                     1.1.2             
            silverstripe/externallinks                 2.0.2             
            silverstripe/framework                     4.1.2             
            silverstripe/graphql                       1.1.2             
            silverstripe/html5                         2.0.1             
            silverstripe/lumberjack                    2.0.0             
            silverstripe/recaptcha                     dev-master 6c839da
            silverstripe/recipe-cms                    1.1.2             
            silverstripe/recipe-core                   1.1.2             
            silverstripe/recipe-plugin                 1.3.0             
            silverstripe/redirectedurls                2.0.0             
            silverstripe/reports                       4.1.2                      
            silverstripe/segment-field                 2.2.1             
            silverstripe/siteconfig                    4.1.2             
            silverstripe/spamprotection                3.0.0             
            silverstripe/tagfield                      2.1.0             
            silverstripe/userforms                     5.2.0             
            silverstripe/vendor-plugin                 1.3.3             
            silverstripe/versioned                     1.1.2             
            swiftmailer/swiftmailer                    v5.4.9            
            symbiote/silverstripe-gridfieldextensions  3.2.0             
            symfony/cache                              v3.4.12           
            symfony/config                             v3.4.12           
            symfony/filesystem                         v4.1.1            
            symfony/finder                             v3.4.12           
            symfony/polyfill-apcu                      v1.8.0            
            symfony/polyfill-ctype                     v1.8.0            
            symfony/polyfill-mbstring                  v1.8.0            
            symfony/translation                        v2.8.42           
            symfony/yaml                               v3.4.12           
            unclecheese/display-logic                  2.0.1             
            undefinedoffset/sortablegridfield          2.0.6             
            webmozart/assert                           1.3.0             
            webonyx/graphql-php                        v0.8.0            
            wilr/silverstripe-googlesitemaps           2.1.2             

When referencing https://github.com/silverstripe/silverstripe-framework/issues/8228 it seems like it may be because we are both using webpack for our front end code, but can't say if that is the cause or not either.  I am in the process of doing a clean install and changing each step to replicate my previous site to see where it breaks.

**Comment by kinglozzer** [comment]
_From @anishpixel on July 19, 2018 9:47_

I am having the same issue, I am using webpack for my web build too so think something is conflicting.

**Comment by kinglozzer** [comment]
_From @spark-green on July 31, 2018 0:15_

Hi guys,
I am having the same issue, 
I am not using  S3 module. but using webpack to run foundation on front-end.

And surprising it is happening only with one single page in CMS.
screenshot 

![image](https://user-images.githubusercontent.com/2790964/43430193-06f07834-94ab-11e8-9b2d-63c09538e8ef.png)


Console error as below
![image](https://user-images.githubusercontent.com/2790964/43430058-6c1b3e3e-94aa-11e8-8018-cb7f74f3160f.png)


For reference I am using below version of silver-stripe and other modules

        

> "php": ">=5.6.0",
>         "silverstripe/cms": "4.1.1",
>         "silverstripe/framework": "4.1.1",
>         "silverstripe/recipe-plugin": "^1",
>         "silverstripe/recipe-cms": "1.*",
>         "silverstripe/versioned": "1.*",
>         "silverstripe/blog": "3.*",
>         "silverstripe/widgets": "2.*",
>         "ryanpotter/silverstripe-cms-theme": "2.*",
>         "undefinedoffset/sortablegridfield": "2.*",
>         "silverstripe/staticpublishqueue": "4.*",
>         "jonom/focuspoint": "dev-master",
>         "unclecheese/display-logic": "2.*",
>         "sheadawson/silverstripe-linkable": "2.*",
>         "silverware/validator": "1.*",
>         "jonom/silverstripe-text-target-length": "2.*",
>         "axllent/silverstripe-cms-tweaks": "2.*",
>         "symbiote/silverstripe-gridfieldextensions": "^3.0"

Any help on this is much appreciated.

**Comment by kinglozzer** [comment]
_From @jpmcc on August 8, 2018 14:19_

Also experiencing similar issue. However, I _am_ using the S3 module, and _not_ using Webpack. The TinyMCE script is being loaded from S3.

Just currently trying to work out why that is placed into assets when all other admin JS isn't

**Comment by kinglozzer** [comment]
_From @anishpixel on August 8, 2018 14:54_

No reply from SS mods or contributors? Seems odd to avoid this as various people are having this issue...

**Comment by kinglozzer** [comment]
_From @jpmcc on August 8, 2018 15:2_

It looks to be a timing thing. The script `tinymce-cms-<unique-string>.js` is  actually loaded from S3, but I am guessing it is after it is actually called. I assume that is why if the page is reloaded, it appears to work as it is now already in the cache. However going back some time later, it is requested again but with a cache busting parameter.

I can see why it is generated as it references the site domain to setup a `baseUrl` var, also based on a specific configuration of the editor.

Just wondering what can be done about that.

Not totally sure if this fits in with everyone else's experience if not using the S3 module.

**Comment by kinglozzer** [comment]
_From @jpmcc on August 8, 2018 15:26_

When reloading, the requirements system places the call for the script into the page source, as you would expect, therefore I assume everything blocks until all the scripts have loaded, and everything works.

Navigating within the CMS, when you move to an area that requires the TinyMCE editor, the JavaScript requirements are pulled in via Ajax and an`x-include-js` header. I assume there is some timing issue here between the script being requested and the editor trying to load, but before the script has loaded in, hence the `tinymce is not defined` error.

Is there any way of forcing the request for the TinyMCE script earlier, or delaying the editor from trying to initialise for a bit longer?

**Comment by kinglozzer** [comment]
_From @jpmcc on August 8, 2018 15:45_

Is there a way to configure the Requirements system to place the TinyMCE generated file to a local folder? There is some information on requirements about[ changing where combined assets are placed](https://docs.silverstripe.org/en/4/developer_guides/templates/requirements/#configuring-combined-file-storage), but so far not making any headway with that.

**Comment by kinglozzer** [comment]
@jpmcc The behaviour of `x-include-js` is the same without the S3 module installed, so provided the request to the `tinymce-cms.js` is returning a valid response I can’t see why it would result in a race condition or anything. It’s not being blocked by CORS headers or something is it?

**Comment by kinglozzer** [comment]
You _might_ be able to override the asset handler for TinyMCE generated files. Totally untested code below, but it might provide a starting point:

```yml
SilverStripe\Core\Injector\Injector:
  League\Flysystem\Filesystem.localpublic:
    class: League\Flysystem\Filesystem
    constructor:
      FilesystemAdapter: '%$SilverStripe\Assets\Flysystem\PublicAssetAdapter'
      FilesystemConfig:
        visibility: public
  LocalGeneratedAssetHandler:
    class: SilverStripe\Assets\Flysystem\GeneratedAssets
    properties:
      Filesystem: '%$League\Flysystem\Filesystem.localpublic'
  SilverStripe\Forms\HTMLEditor\TinyMCECombinedGenerator:
    properties:
      AssetHandler: '%$LocalGeneratedAssetHandler'
```

**Comment by kinglozzer** [comment]
_From @jpmcc on August 9, 2018 8:4_

Thanks Loz, I think I was going down the same path based on the docs, but I think I'm still missing something. The `GeneratedAssetHandler` being injected into the TinyMCE generator still uses the `SilverStripe\S3\Adapter\PublicCacheAdapter` as the Adaptor.

**Comment by kinglozzer** [comment]
_From @jpmcc on August 9, 2018 8:29_

Not sure the configuration is even being picked up :-(. Created a `yml` file in the `_config` folder. I would expect to see an instance of `League\Flysystem\Filesystem` being created. The only one that triggers the constructor is for the `S3` adapter.

Therefore I'm assuming I am missing something pretty fundamental. All my other `yml` files appear to be processed (multiple builds and flushes).

**Comment by kinglozzer** [comment]
Maybe you need a YAML header to ensure it’s loaded last

```yml
---
After:
  - '#corehtml'
---
SilverStripe\Core\Injector\Injector:
  League\Flysystem\Filesystem.localpublic:
# ...
```

**Comment by kinglozzer** [comment]
_From @jpmcc on August 9, 2018 8:33_

yup, using 
```
---
Name: tinymce_override
After:
  - '#silverstripes3-flysystem'
---
```
I'll try `#corehtml`

**Comment by kinglozzer** [comment]
_From @jpmcc on August 9, 2018 8:36_

No luck. 

Really appreciate the help :). I may have to come back to this later.

**Comment by kinglozzer** [comment]
The following does seem to work in my testing - `App\TestAssetAdapter` is the adapter used in `TinyMCECombinedGenerator`:

```yml
---
After:
  - '#corehtml'
---
SilverStripe\Core\Injector\Injector:
  League\Flysystem\Filesystem.localpublic:
    class: 'League\Flysystem\Filesystem'
    constructor:
      FilesystemAdapter: '%$App\TestAssetAdapter'
      FilesystemConfig:
        visibility: public
  LocalGeneratedAssetHandler:
    class: 'SilverStripe\Assets\Flysystem\GeneratedAssets'
    properties:
      Filesystem: '%$League\Flysystem\Filesystem.localpublic'
  SilverStripe\Forms\HTMLEditor\TinyMCECombinedGenerator:
    properties:
      AssetHandler: '%$LocalGeneratedAssetHandler'
```

```php
namespace App;

use SilverStripe\Assets\Flysystem\PublicAssetAdapter;

class TestAssetAdapter extends PublicAssetAdapter
{
}
```

Probably best at this point to hop on the [forums](https://forum.silverstripe.org/) or [Slack](https://www.silverstripe.org/community/slack-signup/) to save adding any more to this issue 😄

**Comment by kinglozzer** [comment]
@jpmcc has confirmed that the above will allow the TinyMCE JS to be written to the local filesystem rather than S3, which offers a workaround for the issue for now.

As this _definitely_ occurs with the S3 module installed, I’m going to move this issue to that repo. Even if it requires a fix in framework, it’s best to track the issue there as that module seems to be the only definite cause.

For those experiencing this issue without S3 installed (@spark-green, @anishpixel?) I’m going to re-open https://github.com/silverstripe/silverstripe-framework/issues/8228 as there’s less S3-related noise there. If you’re still experiencing this issue, please paste the output of `composer show` into that issue so we know which exact versions of each module you’re using. Thanks!

**Comment by kinglozzer** [comment]
Migrated from https://github.com/silverstripe/silverstripe-cms/issues/2177. There’s a workaround available here https://github.com/silverstripe/silverstripe-cms/issues/2177#issuecomment-411688248

**Comment by eehondas** [comment]
Hi @kinglozzer , thanks for opening this.
It is happening consistently for me as well on any given SS4 project.
This is a sample composer extract
`    "require": {
        "php": ">=5.6.0",
        "silverstripe/cms": "^4.2",
        "silverstripe/framework": "^4.2",
        "silverstripe/recipe-plugin": "*",
        "silverstripe/recipe-cms": "*",
        "silverstripe/versioned": "1.*",
        "silverstripe/blog": "3.*",
        "silverstripe/widgets": "2.*",
        "silverstripe/userforms": "5.*",
        "silverstripe/redirectedurls": "2.*",
        "ryanpotter/silverstripe-cms-theme": "2.*",
        "undefinedoffset/sortablegridfield": "2.*",
        "silverstripe/staticpublishqueue": "4.*",
        "silverware/validator": "1.*",
        "unclecheese/display-logic": "2.*",
        "sheadawson/silverstripe-linkable": "2.*",
        "jonom/silverstripe-text-target-length": "2.*",
        "jonom/focuspoint": "3.*",
        "dnadesign/silverstripe-elemental": "3.0.x-dev",
        "axllent/silverstripe-cms-tweaks": "2.*",
        "heyday/silverstripe-menumanager": "~3.0.0",
        "heyday/silverstripe-colorpalette": "2.*"
    }`
Unfortunately your proposed solution does not seem to fix the issue at all for me.

**Comment by kinglozzer** [comment]
@eehondas It doesn't look like you have the s3 module installed. If you’re experiencing that issue without using the s3 module, can you please post here instead: https://github.com/silverstripe/silverstripe-framework/issues/8228? Thanks :)

**Comment by eehondas** [comment]
@kinglozzer Yes.
Done thank you. I just posted here because I followed from the other thread.

**Comment by JoshuaCarter** [comment]
Bump for great justice.

**Comment by madmatt** [comment]
Hey @JoshuaCarter, sorry for the delay - there's not a lot of support for this module as I'm not actively using it in production, and it was only ever meant as a proof-of-concept to test the new SS4 filesystem layer (I'm not entirely sure of the reasons why it was picked up under the `silverstripe` GitHub org :))

Anyway - have you been able to try the workaround that was posted by @kinglozzer here: https://github.com/silverstripe/silverstripe-cms/issues/2177#issuecomment-411688248? I'm keen to know if that still works for you or not, as it might help us assess where the problem lies.

I have a feeling it could be a race condition (TinyMCE file is written to S3 -> attempts to read immediately -> file can't be read -> file gets rewritten next request -> etc) but I haven't been able to confirm that as the cause yet. _Theoretically_ the module shouldn't return an S3 URL unless the file was successfully replicated to S3 and exists in the S3 bucket (e.g. it _should_ wait until the file is replicated within S3 before returning the URL) but perhaps we're missing something there.

**Comment by obj63mc** [comment]
@JoshuaCarter - I submitted a PR that has the workaround for this bug auto included (force tinymce files to write to the local filesystem).  If you checkout the TinyMCEAdapter as well as the .yml file.  If you implement that on your project you should have no issues.

**Comment by madmatt** [comment]
Hey @JoshuaCarter, I've just merged this in, so using the latest `dev-master` should resolve this for you now. Either @obj63mc or I will probably release a new version once #23 is merged.

**Comment by madmatt** [comment]
I've just released [0.4.0](https://github.com/silverstripe/silverstripe-s3/releases/tag/0.4.0) that should hopefully resolve this for you!

Please let us know if you continue to have issues with this.
